### PR TITLE
Cap 292 time series obs data

### PR DESCRIPTION
**Pull request by tomw66**
Added functionality to get timeseries plots for flight data working. 

https://metoffice.atlassian.net/browse/CAP-292?atlOrigin=eyJpIjoiYjU5Y2YyZjg0NWVmNGRlZmEwYzA0YWY4ZGU4OGZlODIiLCJwIjoiaiJ9

### CODE REVIEW

File: src/clean_air/data/data_subset.py [hunk 1]
```diff
@@ -114,16 +113,33 @@ def extract_box(self, box, crs=None):
...
+            df = df.between_time(start, end)
+            if df.empty:
+                raise ValueError('Empty dataframe, likely due to time bounds being out of range')
+
+        series = util.cubes.extract_series(cube, df, column_mapping={'time': 'time'})
+        output_cube = iris.pandas.as_cube(series)
+        output_cube.rename(cube.name())
+        output_cube.units = cube.units
+
+        return output_cube
```
Comment (corinnebosley) [inline src/clean_air/data/data_subset.py:142]: There's a lot of flipping around from one type of dataset to another here, could you just explain what's going on here please?

File: tests/unit/data/test_data_subset.py [hunk 2]
```diff
@@ -90,6 +88,29 @@ def test_as_cube_latlon(dataset):
...
+        # Check we have the correct time bounds (in hours since epoch)
+        assert cube.coord('index').points[0] == int(449197)
+        assert cube.coord('index').points[-1] == int(449198)
+
+    @staticmethod
+    def test_time_bound_error(dataset):
+        with pytest.raises(ValueError, match='Empty dataframe, likely due to time bounds being out of range'):
+            dataset.extract_track('20:00', '21:00')
+
+
```
Comment (corinnebosley) [inline tests/unit/data/test_data_subset.py:113]: Good tests, but I would like to see a description at the top of every test please.

File: tests/unit/data/test_extract_metadata.py [hunk 3]
```diff
@@ -305,14 +305,14 @@ def test_containing_polygon_overlap_within_separate(cube_1, cube_2, cube_3, cube
     @staticmethod
     def test_aircraft_cube(aircraft_cube):
         """
-        GIVEN a single cube
+        GIVEN a cubelist of aircraft data
```
Comment (corinnebosley) [inline tests/unit/data/test_extract_metadata.py:308]: I feel that we should have a unit test of both of these instances; one using a single cube and one using a cubelist.  If we don't have the sample data of this available right now then it is best to write both tests and comment out the one without sample data (with a comment about why this is the case) and write a Jira ticket to reinstate tests once data is available.  Also write a Jira ticket to obtain or create the missing data.  Or put both tasks on same ticket.  Just as long as we don't forget why a test is commented out.

File: tests/unit/util/test_cubes.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    return iris.load_cube(path, 'NO2_concentration_ug_m3')
+
+@pytest.fixture(scope="class")
+def track_dataframe(aircraft_cube):
+	ds = xr.DataArray.from_iris(aircraft_cube)
+	return ds.to_dataframe()
+
+class TestExtract:
+
+	def test_extract_series(self, aircraft_cube, track_dataframe):
```
Comment (corinnebosley) [inline tests/unit/util/test_cubes.py:25]: I would like to see descriptions of class and test in here please, but other than that very good.

**Review by corinnebosley — CHANGES_REQUESTED** [review]
I've left a few comments in the review here; mostly easy but inportant things like adding test descriptions, but I am also a bit baffled by all the dataframe flipping in data_subset.py (I think), so adding some comments to explain what's going on in there would be great too.  thanks.

File: src/clean_air/data/data_subset.py (hunk 1 above)
Comment (tomw66) [reply src/clean_air/data/data_subset.py:142]: Passed to `extract_series` is the original iris cube, and a pandas dataframe version of the data that is cut down to the time period we're interested in. Within `extract_series`, these are used to interpolate the cube along the track, that is then converted and returned as a pandas series. `extract_track` then converts this back to a cube. 

So for this particular case the final switch to a series and back again isn't needed - but it's using a more general helper function that is designed to extract a series from a cube, so I think a good idea to leave that as returning a series.

On the other hand! It does take some time to run; ~1 min for data with 1000 timesteps. So maybe it's a good idea to rethink this process from the ground up, and ignore what was written previously? Slightly difficult to say without knowing other potential use cases for the functions involved.

(Sorry for the stream of consciousness rambling!)

File: tests/unit/data/test_data_subset.py (hunk 2 above)
Comment (tomw66) [reply tests/unit/data/test_data_subset.py:113]: Should we agree between us on a consistent style to use overall? Some have simple 1 line description, a few others use 'GIVEN...WHEN...THEN...' structure. I suppose the latter is better, but can be tricky for some of them.

File: tests/unit/data/test_extract_metadata.py (hunk 3 above)
Comment (tomw66) [reply tests/unit/data/test_extract_metadata.py:308]: The two tests below this cover both of these instances (the function working for a cube and a cubelist); the aim of this test is to ensure it works with the some real flight data. 
Here I've just updated the description after I changed the sample data to cover more than one variable, for the multiplots in the notebook (but is also probably more accurate for testing, as there are no aircraft files that are single cubes).

File: src/clean_air/data/data_subset.py (hunk 1 above)
Comment (corinnebosley) [reply src/clean_air/data/data_subset.py:142]: I'm still wondering what the use case would be in which we need this to be converted to a pandas series and then back again.  Let's chat about it later.

File: tests/unit/data/test_data_subset.py (hunk 2 above)
Comment (corinnebosley) [reply tests/unit/data/test_data_subset.py:113]: I agree.  I think for all new tests we should use the 'GIVEN... WHEN... THEN...' structure, or as close as we can get to it, and if we have time we can convert old tests to the same or similar format.  Still a bit of a nice-to-have though.

File: tests/unit/data/test_extract_metadata.py (hunk 3 above)
Comment (corinnebosley) [reply tests/unit/data/test_extract_metadata.py:308]: Well then I apologise.  Nice work.

**Review by corinnebosley — APPROVED** [review]

**Comment by tomw66** [comment]
1. Refactored `data_subset.extract_track` to only need iris cubes; exact same user functionality but now a lot faster!
2. Removed `cubes.extract_series` as this is no longer used.
3. As such, have removed the `cubes` unit tests as they no longer test anything; `cubes` unit tests are instead now being added in branch **CAP-321**.
4. Improved relevant unit tests, along with descriptions.

**Comment by tomw66** [comment]
@corinnebosley getting an extra failing test now: `ERROR tests/unit/util/test_dataframe_converter.py::TestConvertToGeoDF::test_1d_cube_series` - I assume this is fixed in your branch and can be ignored?

**Review by corinnebosley — CHANGES_REQUESTED** [review]
The changes look great, I'm really excited that we're cutting out the bits we don't need here.  With any luck we can make it much more efficient!

That test is not fixed on my branch though; I'm working on unit tests for dataset_renderer.py, as opposed to dataframe_converter.py.  But that test is failing because I've moved the sample data around.  You can either dig into it and find some data that will work with the failing test, or you can comment it out with a useful description of why we have done this and write a jira ticket to fix this test.  Your call.

**Pull request merged**